builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+    const Twine&   instName)       // [in] Name to give final instruction.
+{
+    if (getContext().GetGfxIpVersion().major >= 8)
+    {
+        Value* const pIdentity = CreateGroupOperationIdentity(groupOperation, pValue->getType());
+
+        // Start the WWM section by setting the inactive invocations.
+        Value* const pSetInactive = CreateSetInactive(pValue, pIdentity);
+
+        Value* pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(2)),
+            CreateGroupOperationArithmetic(groupOperation, pSetInactive,
+                CreateDPPUpdate(pIdentity, pSetInactive, DPPCtrl::DPP_ROW_SR1, 0xf, 0xf, 0)), pSetInactive);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(4)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pSetInactive, DPPCtrl::DPP_ROW_SR2, 0xf, 0xf, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(4)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pSetInactive, DPPCtrl::DPP_ROW_SR3, 0xf, 0xf, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(8)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_SR4, 0xf, 0xe, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(16)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_SR8, 0xf, 0xc, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpUGE(pClusterSize, getInt32(32)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_BCAST15, 0xa, 0xf, 0)), pResult);
+
+        pResult = CreateSelect(CreateICmpEQ(pClusterSize, getInt32(64)),
+            CreateGroupOperationArithmetic(groupOperation, pResult,
+                CreateDPPUpdate(pIdentity, pResult, DPPCtrl::DPP_ROW_BCAST31, 0xc, 0xf, 0)), pResult);
+
+        // Finish the WWM section by calling the intrinsic.
+        return CreateWWM(pResult);
+    }
+    else
+    {
+        Value* const pClusteredExclusive = CreateSubgroupClusteredExclusive(groupOperation,
+                                                                            pValue,
+                                                                            pClusterSize,
+                                                                            instName);
+        return CreateGroupOperationArithmetic(groupOperation, pValue, pClusteredExclusive);
+    }
+}
+
+// =====================================================================================================================
+// Create a subgroup clustered exclusive scan.
+Value* BuilderImplSubgroup::CreateSubgroupClusteredExclusive(
```

Comment:
It seems you do subgroup algorithm operations in vector level, instead of expand it to scalar first.
Is it by design? It seems CreateDPPUpdate and CreateDSSwizzle can't support vector, you have to insert multiple extract elelment and insert element in it. 
If I remember correctly, our original implementation (.ll based) and close source driver always expand the vector to scalar first. I am not sure which style is more friendly for LLVM.

Reply:
I do it at the vector level yup - this produces better code for two reasons:

1. It means we have less enter WWM -> exit WWM regions (less exec math) in the final asm.
1. It means that for things like fp16/int16 where we can do rpm on GFX9 we can now produce it.